spec/index.bs

```diff
@@ -4625,19 +4669,58 @@ interface mixin GPUProgrammablePassEncoder {
 
             **Arguments:**
             <pre class=argumentdef for="GPUProgrammablePassEncoder/setBindGroup(index, bindGroup, dynamicOffsetsData, dynamicOffsetsDataStart, dynamicOffsetsDataLength)">
-                index:
-                bindGroup:
-                dynamicOffsetsData:
-                dynamicOffsetsDataStart:
-                dynamicOffsetsDataLength:
+                |index|: The index to set the bind group at.
+                |bindGroup|: Bind group to use for subsequent render or compute commands.
+                |dynamicOffsetsData|: Array containing buffer offsets in bytes for each entry in
+                    |bindGroup| with marked as {{GPUBindGroupLayoutEntry/hasDynamicOffset}}.
+                |dynamicOffsetsDataStart|: Offset in elements into |dynamicOffsetsData| where the
+                    buffer offset data begins.
+                |dynamicOffsetsDataLength|: Number of buffer offsets to read from |dynamicOffsetsData|.
             </pre>
 
             **Returns:** void
 
-            Issue: Describe {{GPUProgrammablePassEncoder/setBindGroup(index, bindGroup, dynamicOffsetsData, dynamicOffsetsDataStart, dynamicOffsetsDataLength)}} algorithm steps.
+            Issue the following steps on the [=Device timeline=] of |this|.{{GPUObjectBase/[[device]]}}:
+            <div class=device-timeline>
+                1. If any of the following conditions are unsatisfied, make |this| [=invalid=] and stop.
+                    <div class=validusage>
+                        - |bindGroup| is [$valid to use with$] |this|.
+                        - |index| &lt; {{GPULimits/maxBindGroups|GPULimits.maxBindGroups}}.
+                        - |dynamicOffsetsDataLength| is
+                            |bindGroup|.{{GPUBindGroup/[[layout]]}}.{{GPUBindGroupLayout/[[dynamicOffsetCount]]}}.
+                        - |dynamicOffsetsDataStart| + |dynamicOffsetsDataLength| &le; |dynamicOffsetsData|.length.
+
+                        - [$Iterate over each dynamic binding offset$] in |bindGroup| and
+                            run the following steps for each |bufferBinding|, |minBufferBindingSize|,
+                            and |dynamicOffsetIndex|:
+
+                            - Let |bufferDynamicOffset| be
+                                |dynamicOffsetsData|[|dynamicOffsetIndex| + |dynamicOffsetsDataStart|].
+                            - |bufferBinding|.{{GPUBufferBinding/offset}} + |bufferDynamicOffset| +
+                                |minBufferBindingSize| &le;
+                                |bufferBinding|.{{GPUBufferBinding/buffer}}.{{GPUBuffer/[[size]]}}.
+                    </div>
+                1. Set |this|.{{GPUProgrammablePassEncoder/[[bind_groups]]}}[|index|] to be |bindGroup|.
+            </div>
         </div>
 </dl>
 
+<div algorithm>
```

Comment:
Perhaps we should pre-bake the list of dynamically-offset buffer bindings in createBindGroupLayout. I think it would just be a sorted list of binding numbers like `[1, 3, 4]` and you could use those to index back into the bind group (layout) `entries`